File: src/main/java/io/github/faecraft/heartshapedbox/logic/damage/DamageHandlerDispatcher.java
package io.github.faecraft.heartshapedbox.logic.damage;

import io.github.faecraft.heartshapedbox.bad.BadMixinAtomicFlags;
import io.github.faecraft.heartshapedbox.body.BodyPartProvider;
import io.github.faecraft.heartshapedbox.logic.damage.impl.*;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.stat.Stats;
import net.minecraft.util.Pair;

import java.util.ArrayList;

public class DamageHandlerDispatcher {
    private static final ArrayList<DamageHandler> handlers = new ArrayList<>();

    public static boolean handleDamage(ServerPlayerEntity player, DamageSource source, float amount) {
        BodyPartProvider provider = (BodyPartProvider)player;
    
        if (player.hasStatusEffect(StatusEffects.RESISTANCE) && source != DamageSource.OUT_OF_WORLD) {
            //noinspection ConstantConditions
            int k = (player.getStatusEffect(StatusEffects.RESISTANCE).getAmplifier() + 1) * 5;
            int j = 25 - k;
            float f = amount * (float)j;
            float g = amount;
            amount = Math.max(f / 25.0F, 0.0F);
            float h = g - amount;
            if (h > 0.0F && h < 3.4028235E37F) {
                player.increaseStat(Stats.DAMAGE_RESISTED, Math.round(h * 10.0F));
            }
        }
    
        // Done to deal damage in a single go
        float collectedDamage = 0f;
        // Saves state before damage handled
        CompoundTag stateBefore = provider.writeToTag();
        for (DamageHandler possibleHandler : handlers) {
            if (possibleHandler.shouldHandle(source)) {
                Pair<Boolean, Float> result = possibleHandler.handleDamage(player, (BodyPartProvider)player, source, amount);

                collectedDamage += amount - result.getRight();
                amount = result.getRight();

                if (result.getLeft() || result.getRight() <= 0) {
                    break;
                }
            }
        }
        // Try to deal damage, roll back if failed
        // see BadMixinAtomicFlags for the reason behind this
        BadMixinAtomicFlags.callSuperDamage.set(true);
        boolean didDealDamage = player.damage(source, collectedDamage);
        BadMixinAtomicFlags.callSuperDamage.set(false);

        // Revert the state if vanilla doesn't like our damage for whatever reason
        if (!didDealDamage) {
            provider.readFromTag(stateBefore);
        }
        return didDealDamage;
    }

    public static void registerHandlers() {
        // Lower indices get tested first (i.e higher priority)
        handlers.add(new ProjectileDamageHandler());
        handlers.add(new FallingBlockDamageHandler());
        handlers.add(new FallDamageHandler());
        handlers.add(new HotFloorDamageHandler());
        handlers.add(new GenericDamageHandler());
    }
}

File: src/main/kotlin/io/github/faecraft/heartshapedbox/logic/damage/DamageHandlerDispatcher.kt
package io.github.faecraft.heartshapedbox.logic.damage

import io.github.faecraft.heartshapedbox.bad.BadMixinAtomicFlags
import io.github.faecraft.heartshapedbox.body.BodyPartProvider
import io.github.faecraft.heartshapedbox.logic.damage.impl.*
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.DamageUtil
import net.minecraft.entity.damage.DamageSource
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.stat.Stats
import java.util.*
import kotlin.math.max
import kotlin.math.roundToInt

public object DamageHandlerDispatcher {
    private val handlers = ArrayList<DamageHandler>()

    @JvmStatic
    @Suppress("MagicNumber")  // Based on vanilla code
    public fun handleDamage(player: ServerPlayerEntity, source: DamageSource, amount: Float): Boolean {
        var mutAmount = amount
        val provider = player as BodyPartProvider

        // Taken directly from damage logic
        if (player.hasStatusEffect(StatusEffects.RESISTANCE) && source != DamageSource.OUT_OF_WORLD) {
            val k = (player.getStatusEffect(StatusEffects.RESISTANCE)!!.amplifier + 1) * 5
            val j = 25 - k
            val f = mutAmount * j.toFloat()
            val g = mutAmount
            mutAmount = max(f / 25.0f, 0f)
            val h = g - mutAmount
            if (h > 0.0f && h < Float.MAX_VALUE) {
                player.increaseStat(Stats.DAMAGE_RESISTED, (h * 10.0f).roundToInt())
            }
        }

        // Done to deal damage in a single go
        var collectedDamage = 0f
        // Saves state before damage handled
        val stateBefore = provider.writeToTag()
        for (possibleHandler in handlers) {
            if (possibleHandler.shouldHandle(source)) {
                val items = possibleHandler.getPossibleArmorPieces(source, player)
                val k = EnchantmentHelper.getProtectionAmount(items, source)

                if (k > 0) {
                    mutAmount = DamageUtil.getInflictedDamage(mutAmount, k.toFloat())
                }

                val result = possibleHandler.handleDamage(player, player as BodyPartProvider, source, mutAmount)

                collectedDamage += mutAmount - result.second
                mutAmount = result.second

                if (result.first || result.second <= 0) {
                    break
                }
            }
        }

        // Try to deal damage, roll back if failed
        // see BadMixinAtomicFlags for the reason behind this
        BadMixinAtomicFlags.callSuperDamage.set(true)
        val didDealDamage = player.damage(source, collectedDamage)
        BadMixinAtomicFlags.callSuperDamage.set(false)

        // Revert the state if vanilla doesn't like our damage for whatever reason
        if (!didDealDamage) {
            provider.readFromTag(stateBefore)
        }

        return didDealDamage
    }

    public fun registerHandlers() {
        // Lower indices get tested first (i.e higher priority)

        // Likely entities
        handlers.add(EntityDamageHandler())
        handlers.add(ProjectileDamageHandler())

        // Environment
        handlers.add(DrownDamageHandler())
        handlers.add(FallingBlockDamageHandler())
        handlers.add(FallDamageHandler())
        handlers.add(HotFloorDamageHandler())

        // Other
        handlers.add(GenericAndRemainingDamageHandler())
    }
}
